src/libsyntax_ext/concat.rs

```diff
@@ -62,5 +83,32 @@ pub fn expand_syntax_ext(
         err.emit();
     }
     let sp = sp.apply_mark(cx.current_expansion.mark);
-    base::MacEager::expr(cx.expr_str(sp, Symbol::intern(&accumulator)))
+    // Do not allow mixing "" and b"", but return the joint b"" to avoid further errors
+    if string_accumulator.len() > 0 && b_accumulator.len() > 0 {
+        let mut err = cx.struct_span_err(
+            b_pos.clone(),
+            "cannot concatenate a byte string literal with string literals",
+        );
+        for pos in &b_pos {
+            err.span_label(*pos, "byte string literal");
+        }
+        for pos in &string_pos {
+            err.span_label(*pos, "string literal");
+
+        }
+        err.help("do not mix byte string literals and string literals");
+        err.multipart_suggestion(
+            "you can use byte string literals",
+            string_pos
+                .iter()
+                .map(|pos| (pos.shrink_to_lo(), "b".to_string()))
+                .collect(),
+        );
+        err.emit();
```

Comment:
Just noticed that this error (and suggestion) will be wrong when doing `concat!(b"asdf", true, 3)`.